### ISSUE TITLE
Make non-battery saver mode more efficient

### ISSUE THREAD

**Issue by sarav** [farkam135/GoIV#300]
I've been searching online the past few days for ways to detect any touch on the screen (without caring/getting the info where exactly the touch was -- I'm sure this won't be possible due to security concerns) and there are various suggestions. I'm not which of them necessarily work. If anyone here knows the answer to this already or wants to write some test app to see if this can be done, it'll be good to hear the results.

**IF** we can know that a touch actually happened (I think someone else mentioned it here or on reddit) then we can have the screen scan for pokemon screen to happen only during touches. We'll still have to throttle touch based scanning since I'm sure people tap way faster than the every .75 second scan we are doing now. But that should be easy to do.

I'm hoping that if we don't acquire images from the mediaprojection and let it fill up the 2 image queue in the ImageReader, that the GPU or whatever will stop filling in the image (which I think is the most power hungry part). But maybe not -- because then "acquireLatestImage()" won't be latest anymore. In any case, even if the GPU/whatever is filling up the buffer with the screen content, at least not reading it and checking it every 0.75 seconds should help quite a bit too.

Even though I want to get screenshot/battery saver mode to work, the normal mode is a lot more convenient. So, I'd ideally like to reduce it's power consumption to a point that the benefits outweigh the need to use battery saver mode (we'll still leave it in).

Long story short, anyone willing to figure out how to detect screen touches? Well, even better would be to detect the Pokemon screen without doing any screenscan, but that doesn't look like it's feasible.

**Comment by farkam135** [comment]
Ah interesting, What if the pokefly service displayed a full screen transparent overlay. If we put an onclick on it, shouldnt that act as a way to detect screen touches?

**Comment by sarav** [comment]
AFAIK, I don't think you can let the touch go thru. But I could be wrong.

**Comment by sarav** [comment]
I mean, AFAIK, you can't detect the touch and let it go thru. It might be either or (feels like a security issue otherwise). But maybe accessibility permissions does?

**Comment by rpgrca** [comment]
You mean, something similar to the AMOLED pixel off applications in the Play Store, which display an overlay on top of the screen? I'm guessing it detects clicks and passes them through, but my experience while playing Ingress wasn't that good, the game became clunky since some long pressed were cut, some slides were not detected, etc.

**Comment by sarav** [comment]
I don't think the pixel off or filter applications process any touch. Apparently views can mark themselves as not focusable and won't get any events. 

I don't think lag should be an issue if implemented correctly.

**Comment by sarav** [comment]
Looks like there's a way to do this! This is going to be my next effort. This should hopefully improve power consumption by a lot.

**Comment by Blaisorblade** [comment]
Question that's been nagging me: is the timer we use suspended when the app is? I'd trust Android-specific timers to do that correctly, not sure for ours.

**Comment by sarav** [comment]
@Blaisorblade The timer code is gone now 😃 

@farkam135 A full screen transparent overlay will grab all the touch events and won't let it pass through. There's a specific method to put a view and then say you want to receive events about touches outside the view. For security reasons, you only get the first "finger down" and no location information. This is to avoid apps capturing keyboard key presses, etc to find your password, credit card, etc. Anyway, it seems to work pretty well now.

**Comment by voltaek** [comment]
@sarav Any possibility of this being combined with #263 after you have this in place and working? Basically use the quick screen scan this issue would use, but only do said scan when the floating IV button would be pressed, thus scanning the screen at little as possible (even more so than this issue's method).

**Comment by sarav** [comment]
Well, I wouldn't really call it "combining" this with #263. I completely understand what you are asking in #263 and it should be doable (as in grab the screen only when the floating button is pressed). But the problem is that I'm not a floating button or an UI expert by any means.

Also, you know how the FB messenger floating chat is super smooth when flicking it around? Yeah, that's super hard to achieve -- like deciding the acceleration of the animation when you flick it, how it kinda sticks to the edges when you pull it, etc.

If someone can do the floating button UI, the rest of your request is super easy to do.

**Comment by thearaks** [comment]
@sarav I did it once, for an app called Veer. You can test it and tell me what do you think about it.
There's a lot of work to obtain that floating button motion, but if I can find the needed spare time I think I'm able to implement it...

**Comment by sarav** [comment]
@thearaks Now I'm confused. What exactly do you need from me then? Help with the stuff to do after the button is pressed?

**Comment by sarav** [comment]
@thearaks -- there's a discord channel where devs hangout if you want to chat. We are all different time zones, so you might not be able to catch me all the time.

**Comment by nahojjjen** [comment]
From what I can tell, this was implemented, but the issue was never closed.
